<a href="https://colab.research.google.com/github/Dat2784/Homelander/blob/main/docs/notebooks/Ames_Housing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Select categorical columns
categorical_cols = df.select_dtypes(include=['object', 'category']).columns

# Impute missing values in categorical columns with the mode
for col in categorical_cols:
    if df[col].isnull().any():
        mode_value = df[col].mode()[0]
        df[col].fillna(mode_value, inplace=True)

# Verify that missing values have been imputed
missing_values_after_categorical_imputation = df[categorical_cols].isnull().sum()
display(missing_values_after_categorical_imputation[missing_values_after_categorical_imputation > 0])

/tmp/ipython-input-2927222647.py:8: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna(mode_value, inplace=True)


Series([], dtype: int64)

In [ ]:
# Select numerical columns
numerical_cols = df.select_dtypes(include=['number']).columns

# Impute missing values in numerical columns with the mean
for col in numerical_cols:
    if df[col].isnull().any():
        mean_value = df[col].mean()
        df[col].fillna(mean_value, inplace=True)

# Verify that missing values have been imputed
missing_values_after_imputation = df[numerical_cols].isnull().sum()
display(missing_values_after_imputation[missing_values_after_imputation > 0])

/tmp/ipython-input-3753440199.py:8: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna(mean_value, inplace=True)


Series([], dtype: int64)

In [ ]:
missing_values = df.isnull().sum()
# Display columns with missing values
display(missing_values[missing_values > 0])

Lot Frontage       490
Alley             2732
Mas Vnr Type      1775
Mas Vnr Area        23
Bsmt Qual           80
Bsmt Cond           80
Bsmt Exposure       83
BsmtFin Type 1      80
BsmtFin SF 1         1
BsmtFin Type 2      81
BsmtFin SF 2         1
Bsmt Unf SF          1
Total Bsmt SF        1
Electrical           1
Bsmt Full Bath       2
Bsmt Half Bath       2
Fireplace Qu      1422
Garage Type        157
Garage Yr Blt      159
Garage Finish      159
Garage Cars          1
Garage Area          1
Garage Qual        159
Garage Cond        159
Pool QC           2917
Fence             2358
Misc Feature      2824
dtype: int64

In [ ]:
import pandas as pd

df = pd.read_csv('/content/AmesHousing.csv')
display(df.head())

Order        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
0      1  526301100           20        RL         141.0     31770   Pave   
1      2  526350040           20        RH          80.0     11622   Pave   
2      3  526351010           20        RL          81.0     14267   Pave   
3      4  526353030           20        RL          93.0     11160   Pave   
4      5  527105010           60        RL          74.0     13830   Pave   

  Alley Lot Shape Land Contour  ... Pool Area Pool QC  Fence Misc Feature  \
0   NaN       IR1          Lvl  ...         0     NaN    NaN          NaN   
1   NaN       Reg          Lvl  ...         0     NaN  MnPrv          NaN   
2   NaN       IR1          Lvl  ...         0     NaN    NaN         Gar2   
3   NaN       Reg          Lvl  ...         0     NaN    NaN          NaN   
4   NaN       IR1          Lvl  ...         0     NaN  MnPrv          NaN   

  Misc Val Mo Sold Yr Sold Sale Type  Sale Condition  SalePrice  
0        0       5    2010       WD           Normal     215000  
1        0       6    2010       WD           Normal     105000  
2    12500       6    2010       WD           Normal     172000  
3        0       4    2010       WD           Normal     244000  
4        0       3    2010       WD           Normal     189900  

[5 rows x 82 columns]

# Task
Apply One-Hot Encoding to categorical features and StandardScaler to numerical features in the dataset "/content/AmesHousing.csv".

## Separate features and target

### Subtask:
Separate the target variable (SalePrice) from the features.


**Reasoning**:
Separate the target variable 'SalePrice' from the features in the DataFrame.



In [ ]:
y = df['SalePrice']
X = df.drop('SalePrice', axis=1)

display(y.head())
display(X.head())

0    215000
1    105000
2    172000
3    244000
4    189900
Name: SalePrice, dtype: int64

Order        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
0      1  526301100           20        RL         141.0     31770   Pave   
1      2  526350040           20        RH          80.0     11622   Pave   
2      3  526351010           20        RL          81.0     14267   Pave   
3      4  526353030           20        RL          93.0     11160   Pave   
4      5  527105010           60        RL          74.0     13830   Pave   

  Alley Lot Shape Land Contour  ... Screen Porch Pool Area Pool QC  Fence  \
0  Grvl       IR1          Lvl  ...            0         0      Ex  MnPrv   
1  Grvl       Reg          Lvl  ...          120         0      Ex  MnPrv   
2  Grvl       IR1          Lvl  ...            0         0      Ex  MnPrv   
3  Grvl       Reg          Lvl  ...            0         0      Ex  MnPrv   
4  Grvl       IR1          Lvl  ...            0         0      Ex  MnPrv   

  Misc Feature Misc Val Mo Sold Yr Sold  Sale Type  Sale Condition  
0         Shed        0       5    2010        WD           Normal  
1         Shed        0       6    2010        WD           Normal  
2         Gar2    12500       6    2010        WD           Normal  
3         Shed        0       4    2010        WD           Normal  
4         Shed        0       3    2010        WD           Normal  

[5 rows x 81 columns]

## Identify categorical and numerical features

### Subtask:
Identify which columns are categorical and which are numerical.


**Reasoning**:
Identify the categorical and numerical columns from the features DataFrame X.



In [ ]:
categorical_cols = X.select_dtypes(include=['object', 'category']).columns.tolist()
numerical_cols = X.select_dtypes(include=['number']).columns.tolist()

print("Categorical Columns:")
print(categorical_cols)
print("\nNumerical Columns:")
print(numerical_cols)

Categorical Columns:
['MS Zoning', 'Street', 'Alley', 'Lot Shape', 'Land Contour', 'Utilities', 'Lot Config', 'Land Slope', 'Neighborhood', 'Condition 1', 'Condition 2', 'Bldg Type', 'House Style', 'Roof Style', 'Roof Matl', 'Exterior 1st', 'Exterior 2nd', 'Mas Vnr Type', 'Exter Qual', 'Exter Cond', 'Foundation', 'Bsmt Qual', 'Bsmt Cond', 'Bsmt Exposure', 'BsmtFin Type 1', 'BsmtFin Type 2', 'Heating', 'Heating QC', 'Central Air', 'Electrical', 'Kitchen Qual', 'Functional', 'Fireplace Qu', 'Garage Type', 'Garage Finish', 'Garage Qual', 'Garage Cond', 'Paved Drive', 'Pool QC', 'Fence', 'Misc Feature', 'Sale Type', 'Sale Condition']

Numerical Columns:
['Order', 'PID', 'MS SubClass', 'Lot Frontage', 'Lot Area', 'Overall Qual', 'Overall Cond', 'Year Built', 'Year Remod/Add', 'Mas Vnr Area', 'BsmtFin SF 1', 'BsmtFin SF 2', 'Bsmt Unf SF', 'Total Bsmt SF', '1st Flr SF', '2nd Flr SF', 'Low Qual Fin SF', 'Gr Liv Area', 'Bsmt Full Bath', 'Bsmt Half Bath', 'Full Bath', 'Half Bath', 'Bedroom AbvGr

## Apply one-hot encoding

### Subtask:
Apply One-Hot Encoding to the categorical features.


**Reasoning**:
Apply One-Hot Encoding to the categorical features in the DataFrame X.



In [ ]:
from sklearn.preprocessing import OneHotEncoder

# Instantiate OneHotEncoder
one_hot_encoder = OneHotEncoder(handle_unknown='ignore', sparse_output=False)

# Fit and transform the categorical columns
X_categorical_encoded = one_hot_encoder.fit_transform(X[categorical_cols])

# Convert to DataFrame
X_categorical_encoded_df = pd.DataFrame(X_categorical_encoded, columns=one_hot_encoder.get_feature_names_out(categorical_cols))

display(X_categorical_encoded_df.head())

MS Zoning_A (agr)  MS Zoning_C (all)  MS Zoning_FV  MS Zoning_I (all)  \
0                0.0                0.0           0.0                0.0   
1                0.0                0.0           0.0                0.0   
2                0.0                0.0           0.0                0.0   
3                0.0                0.0           0.0                0.0   
4                0.0                0.0           0.0                0.0   

   MS Zoning_RH  MS Zoning_RL  MS Zoning_RM  Street_Grvl  Street_Pave  \
0           0.0           1.0           0.0          0.0          1.0   
1           1.0           0.0           0.0          0.0          1.0   
2           0.0           1.0           0.0          0.0          1.0   
3           0.0           1.0           0.0          0.0          1.0   
4           0.0           1.0           0.0          0.0          1.0   

   Alley_Grvl  ...  Sale Type_New  Sale Type_Oth  Sale Type_VWD  \
0         1.0  ...            0.0            0.0            0.0   
1         1.0  ...            0.0            0.0            0.0   
2         1.0  ...            0.0            0.0            0.0   
3         1.0  ...            0.0            0.0            0.0   
4         1.0  ...            0.0            0.0            0.0   

   Sale Type_WD   Sale Condition_Abnorml  Sale Condition_AdjLand  \
0            1.0                     0.0                     0.0   
1            1.0                     0.0                     0.0   
2            1.0                     0.0                     0.0   
3            1.0                     0.0                     0.0   
4            1.0                     0.0                     0.0   

   Sale Condition_Alloca  Sale Condition_Family  Sale Condition_Normal  \
0                    0.0                    0.0                    1.0   
1                    0.0                    0.0                    1.0   
2                    0.0                    0.0                    1.0   
3                    0.0                    0.0                    1.0   
4                    0.0                    0.0                    1.0   

   Sale Condition_Partial  
0                     0.0  
1                     0.0  
2                     0.0  
3                     0.0  
4                     0.0  

[5 rows x 267 columns]

## Apply standardscaler

### Subtask:
Apply StandardScaler to the numerical features.


**Reasoning**:
Instantiate StandardScaler and apply it to the numerical features. Then convert the result to a DataFrame.



In [ ]:
from sklearn.preprocessing import StandardScaler

# Instantiate StandardScaler
scaler = StandardScaler()

# Fit and transform the numerical columns
X_numerical_scaled = scaler.fit_transform(X[numerical_cols])

# Convert to DataFrame
X_numerical_scaled_df = pd.DataFrame(X_numerical_scaled, columns=numerical_cols)

display(X_numerical_scaled_df.head())

Order       PID  MS SubClass  Lot Frontage  Lot Area  Overall Qual  \
0 -1.731460 -0.997164    -0.877005      3.366911  2.744381     -0.067254   
1 -1.730277 -0.996904    -0.877005      0.505463  0.187097     -0.776079   
2 -1.729095 -0.996899    -0.877005      0.552372  0.522814     -0.067254   
3 -1.727913 -0.996888    -0.877005      1.115280  0.128458      0.641571   
4 -1.726731 -0.992903     0.061285      0.224010  0.467348     -0.776079   

   Overall Cond  Year Built  Year Remod/Add  Mas Vnr Area  ...  Garage Area  \
0     -0.506718   -0.375537       -1.163488      0.056639  ...     0.256684   
1      0.393091   -0.342468       -1.115542     -0.571242  ...     1.196337   
2      0.393091   -0.441674       -1.259380      0.034215  ...    -0.748092   
3     -0.506718   -0.110988       -0.779919     -0.571242  ...     0.228774   
4     -0.506718    0.848000        0.658466     -0.571242  ...     0.042704   

   Wood Deck SF  Open Porch SF  Enclosed Porch  3Ssn Porch  Screen Porch  \
0      0.920121       0.214409       -0.358838   -0.103134     -0.285354   
1      0.366061      -0.704493       -0.358838   -0.103134      1.854530   
2      2.368594      -0.170937       -0.358838   -0.103134     -0.285354   
3     -0.742060      -0.704493       -0.358838   -0.103134     -0.285354   
4      0.935952      -0.200579       -0.358838   -0.103134     -0.285354   

   Pool Area   Misc Val   Mo Sold   Yr Sold  
0  -0.063031  -0.089422 -0.448057  1.678499  
1  -0.063031  -0.089422 -0.079602  1.678499  
2  -0.063031  21.985725 -0.079602  1.678499  
3  -0.063031  -0.089422 -0.816513  1.678499  
4  -0.063031  -0.089422 -1.184969  1.678499  

[5 rows x 38 columns]

## Combine processed features

### Subtask:
Combine the One-Hot Encoded categorical features and the scaled numerical features into a single DataFrame.


**Reasoning**:
Combine the One-Hot Encoded categorical features and the scaled numerical features into a single DataFrame.



In [ ]:
X_processed = pd.concat([X_categorical_encoded_df, X_numerical_scaled_df], axis=1)
display(X_processed.head())

MS Zoning_A (agr)  MS Zoning_C (all)  MS Zoning_FV  MS Zoning_I (all)  \
0                0.0                0.0           0.0                0.0   
1                0.0                0.0           0.0                0.0   
2                0.0                0.0           0.0                0.0   
3                0.0                0.0           0.0                0.0   
4                0.0                0.0           0.0                0.0   

   MS Zoning_RH  MS Zoning_RL  MS Zoning_RM  Street_Grvl  Street_Pave  \
0           0.0           1.0           0.0          0.0          1.0   
1           1.0           0.0           0.0          0.0          1.0   
2           0.0           1.0           0.0          0.0          1.0   
3           0.0           1.0           0.0          0.0          1.0   
4           0.0           1.0           0.0          0.0          1.0   

   Alley_Grvl  ...  Garage Area  Wood Deck SF  Open Porch SF  Enclosed Porch  \
0         1.0  ...     0.256684      0.920121       0.214409       -0.358838   
1         1.0  ...     1.196337      0.366061      -0.704493       -0.358838   
2         1.0  ...    -0.748092      2.368594      -0.170937       -0.358838   
3         1.0  ...     0.228774     -0.742060      -0.704493       -0.358838   
4         1.0  ...     0.042704      0.935952      -0.200579       -0.358838   

   3Ssn Porch  Screen Porch  Pool Area   Misc Val   Mo Sold   Yr Sold  
0   -0.103134     -0.285354  -0.063031  -0.089422 -0.448057  1.678499  
1   -0.103134      1.854530  -0.063031  -0.089422 -0.079602  1.678499  
2   -0.103134     -0.285354  -0.063031  21.985725 -0.079602  1.678499  
3   -0.103134     -0.285354  -0.063031  -0.089422 -0.816513  1.678499  
4   -0.103134     -0.285354  -0.063031  -0.089422 -1.184969  1.678499  

[5 rows x 305 columns]

## Summary:

### Data Analysis Key Findings

*   The target variable 'SalePrice' was successfully separated from the features.
*   43 categorical columns and 38 numerical columns were identified in the dataset.
*   One-Hot Encoding was successfully applied to the categorical features, resulting in a DataFrame with a larger number of columns representing the different categories.
*   StandardScaler was successfully applied to the numerical features, resulting in a DataFrame where numerical values are scaled.
*   The processed categorical and numerical features were successfully combined into a single DataFrame `X_processed`.

### Insights or Next Steps

*   The combined `X_processed` DataFrame is now ready for use in machine learning models that require numerical input and handle categorical features via One-Hot Encoding.
*   Further analysis could involve checking for missing values in the combined DataFrame and deciding on an imputation strategy if necessary before model training.


# Task
Build a model to predict house sale prices (`SalePrice`) based on land, house, area, and utility attributes using the data in "/content/AmesHousing.csv". The model should be trained and evaluated, and the final output should include the model's performance metrics. Use the information in "/content/Thông tin các thuộc tính.txt" and "/content/Biến đặc trưng.txt" to understand the data and features.

## Split data

### Subtask:
Split the preprocessed data into training and testing sets.


**Reasoning**:
Split the preprocessed data into training and testing sets and display their shapes.



In [ ]:
from sklearn.model_selection import train_test_split

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_processed, y, test_size=0.2, random_state=42)

# Display the shapes of the resulting sets
print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (2344, 305)
Shape of X_test: (586, 305)
Shape of y_train: (2344,)
Shape of y_test: (586,)


## Choose a model

### Subtask:
Select an appropriate machine learning model for regression.


**Reasoning**:
Import and instantiate a RandomForestRegressor model as it is a suitable regression model for this task and generally performs well on various datasets.



In [ ]:
from sklearn.ensemble import RandomForestRegressor

# Instantiate RandomForestRegressor
model = RandomForestRegressor(random_state=42)

## Train the model

### Subtask:
Train the selected regression model using the training data.


**Reasoning**:
Fit the RandomForestRegressor model to the training data.



In [ ]:
# Fit the model to the training data
model.fit(X_train, y_train)

RandomForestRegressor(random_state=42)

## Evaluate the model

### Subtask:
Evaluate the trained model's performance using the testing data.


**Reasoning**:
Evaluate the trained model's performance using the testing data by calculating MSE and R-squared.



In [ ]:
from sklearn.metrics import mean_squared_error, r2_score

# Make predictions on the testing data
y_pred = model.predict(X_test)

# Calculate Mean Squared Error (MSE)
mse = mean_squared_error(y_test, y_pred)

# Calculate R-squared score
r2 = r2_score(y_test, y_pred)

# Print the performance metrics
print(f"Mean Squared Error (MSE): {mse}")
print(f"R-squared (R2) Score: {r2}")

Mean Squared Error (MSE): 697218861.6660963
R-squared (R2) Score: 0.9130383918282162


## Make predictions

### Subtask:
Use the trained model to make predictions on new data.


**Reasoning**:
Use the trained model to make predictions on the test set and display the first few predictions.



In [ ]:
# Make predictions on the testing data
y_pred = model.predict(X_test)

# Display the first few predictions
display(y_pred[:5])

array([187747.5 , 104264.  , 194549.33, 118596.33, 111178.  ])

## Summary:

### Data Analysis Key Findings

*   The data was successfully split into training (2344 samples) and testing (586 samples) sets, with a test size of 0.2.
*   A `RandomForestRegressor` model was selected and trained on the training data.
*   The trained model achieved a Mean Squared Error (MSE) of approximately 697,218,861.67 on the testing data.
*   The trained model achieved an R-squared (\(R^2\)) score of approximately 0.913 on the testing data, indicating that the model explains about 91.3% of the variance in the house sale prices.
*   The model was successfully used to make predictions on the testing data.

### Insights or Next Steps

*   The high \(R^2\) score suggests that the `RandomForestRegressor` is a suitable model for this prediction task. Further investigation into feature importance from the trained model could provide insights into which attributes are the most influential in predicting house sale prices.
*   While the \(R^2\) score is high, the MSE is also large, which is expected given the scale of the target variable (`SalePrice`). Converting the target variable to a logarithmic scale before training could potentially improve the model's performance and provide a more intuitive interpretation of the error metric (e.g., using Root Mean Squared Logarithmic Error).
